portable/NetworkInterface/TM4C/NetworkInterface.c

```diff
@@ -0,0 +1,599 @@
+                           EMAC_FRMFILTER_PASS_MULTICAST));
+
+        // Set the MAC address
+        MAP_EMACAddrSet(EMAC0_BASE, 0, &mac_address_bytes[0]);
+
+        // Clears any previously asserted interrupts
+        MAP_EMACIntClear(EMAC0_BASE, EMACIntStatus(EMAC0_BASE, false));
+
+        // Initialize the DMA descriptors
+        _dma_descriptors_init();
+
+        // Enable TX/RX
+        MAP_EMACTxEnable(EMAC0_BASE);
+        MAP_EMACRxEnable(EMAC0_BASE);
+
+        // Set the interrupt to a lower priority than the OS scheduler interrupts
+        MAP_IntPrioritySet(INT_EMAC0,  (6 << (8 - configPRIO_BITS)));
+
+        // Enable the Ethernet RX and TX interrupt source.
+        MAP_EMACIntEnable(EMAC0_BASE, (EMAC_INT_RECEIVE | EMAC_INT_TRANSMIT |
+                      EMAC_INT_TX_STOPPED | EMAC_INT_RX_NO_BUFFER |
+                      EMAC_INT_RX_STOPPED | EMAC_INT_PHY));
+
+        // Enable EMAC interrupts
+        MAP_IntEnable(INT_EMAC0);
+
+        // Enable all processor interrupts.
+        MAP_IntMasterEnable();
+
+        // Tell the PHY to start an auto-negotiation cycle.
+        MAP_EMACPHYWrite(EMAC0_BASE, PHY_PHYS_ADDR, EPHY_BMCR, (EPHY_BMCR_ANEN |
+                EPHY_BMCR_RESTARTAN));
+    }
+
+    return success;
+}
+
+BaseType_t xNetworkInterfaceOutput(NetworkBufferDescriptor_t * const pxNetworkBuffer, BaseType_t xReleaseAfterSend)
+{
+    BaseType_t success = pdTRUE;
+    tEMACDMADescriptor *dma_descriptor;
+
+    dma_descriptor = &_tx_descriptors[_tx_descriptor_list.write];
+
+    // If the DMA controller still owns the descriptor, all DMA descriptors are in use, bail out
+    if (0 == (dma_descriptor->ui32CtrlStatus & DES0_RX_CTRL_OWN))
+    {
+        // Assign the buffer to the DMA descriptor
+        dma_descriptor->pvBuffer1 = pxNetworkBuffer->pucEthernetBuffer;
+
+        // Inform the DMA of the size of the packet
+        dma_descriptor->ui32Count = (pxNetworkBuffer->xDataLength & DES1_TX_CTRL_BUFF1_SIZE_M) << DES1_TX_CTRL_BUFF1_SIZE_S;
+
+        // Inform the DMA that this is the first and last segment of the packet, calculate the checksums, the descriptors are
+        // chained, and to use interrupts
+        dma_descriptor->ui32CtrlStatus = DES0_TX_CTRL_FIRST_SEG | /*DES0_TX_CTRL_IP_ALL_CKHSUMS |*/ DES0_TX_CTRL_CHAINED
+                | DES0_TX_CTRL_LAST_SEG | DES0_TX_CTRL_INTERRUPT;// | DES0_TX_CTRL_REPLACE_CRC;
+
+        // Advance the index in the list
+        _tx_descriptor_list.write++;
+
+        // Wrap around if required
+        if (_tx_descriptor_list.write == SYSCONFIG_FREERTOS_TCP_TX_DESC_COUNT)
+        {
+            _tx_descriptor_list.write = 0;
+        }
+
+        // Give the DMA descriptor to the DMA controller
+        dma_descriptor->ui32CtrlStatus |= DES0_TX_CTRL_OWN;
+
+        // Inform the DMA it has a new descriptor
+        MAP_EMACTxDMAPollDemand(EMAC0_BASE);
+
+        iptraceNETWORK_INTERFACE_TRANSMIT();
+    }
+    else
+    {
+        success = pdFALSE;
```

Comment:
At this point, sending of a packet has failed. Remember the release the network buffer:
~~~diff
     else
     {
         success = pdFALSE;
+        vReleaseNetworkBufferAndDescriptor( pxNetworkBuffer );
     }
     return success;
~~~
otherwise you'd get a leak.

Reply:
Whoops! Thanks!

Reply:
I see lots of changes, thanks for that!

Also you mentioned that : 

> From the porting guide, it seemed to be implied that it was more advantageous to handle RX packets in the ISR

Oops, I was not aware of that. Can you give me the URL of the page that says so?

And you are right about the conversion function `pxPacketBuffer_to_NetworkBuffer()`, it should be mentioned when talking about zero-copy method.

Reply:
I see, you have read that well. At the same time, the comment says that "The deferred interrupt handler is a standard RTOS task".

I will propose to change that text.

My feeling is that the effect of dropping packets during the ISR will won't have any noticeable effect on the performance. The CPU load will be the same, it's just that the packet will be dropped a fraction of a uS later.
Note that the unwanted packet kept the LAN busy for about 120 µS before it reached our device ( full-size packet on a 100 Mbps LAN ).

I will download your adapted sources and have a new look again.
Thanks a lot so far!